src/solver/sparsesoftmaxcrossentropywithlogits/forward_sparsesoftmaxcrossentropywithlogits.cpp

```diff
@@ -0,0 +1,140 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/datatype.hpp>
+#include <miopen/sparsesoftmaxcrossentropywithlogits.hpp>
+#include <miopen/sparsesoftmaxcrossentropywithlogits/invoke_params.hpp>
+#include <miopen/sparsesoftmaxcrossentropywithlogits/solvers.hpp>
+#include <miopen/mlo_internal.hpp>
+#include <miopen/target_properties.hpp>
+#include <miopen/tensor_view_utils.hpp>
+
+#define LOCAL_SIZE_FWD 256
+
+namespace miopen {
+
+namespace solver {
+
+namespace sparsesoftmaxcrossentropywithlogits {
+
+bool SparseSoftmaxCrossEntropyWithLogitsForward::IsApplicable(
+    const ExecutionContext&,
+    const miopen::sparsesoftmaxcrossentropywithlogits::FwdProblemDescription& problem) const
+{
+    if(!(problem.GetOutputDesc().GetType() == miopenHalf ||
+         problem.GetOutputDesc().GetType() == miopenFloat ||
+         problem.GetOutputDesc().GetType() == miopenBFloat16))
+    {
+        return false;
+    }
+    return true;
+}
+
+ConvSolution SparseSoftmaxCrossEntropyWithLogitsForward::GetSolution(
+    const ExecutionContext& context,
+    const miopen::sparsesoftmaxcrossentropywithlogits::FwdProblemDescription& problem) const
+{
+    std::ignore       = context;
+    auto output_dtype = miopen::GetDataType(problem.GetOutputDesc().GetType());
+    auto target_dtype = miopen::GetDataType(problem.GetTargetDesc().GetType());
+    auto dtype        = problem.GetOutputDesc().GetType();
+    auto dtype_lowest = std::numeric_limits<float>::lowest();
+
+    auto result       = ConvSolution{miopenStatusSuccess};
+    auto build_params = KernelBuildParameters{
+        {"MIOPEN_USE_FP16", static_cast<int>(dtype == miopenHalf)},
+        {"MIOPEN_USE_FP32", static_cast<int>(dtype == miopenFloat)},
+        {"MIOPEN_USE_FP64", static_cast<int>(dtype == miopenDouble)},
+        {"MIOPEN_USE_BFP16", static_cast<int>(dtype == miopenBFloat16)},
+        {"D_TYPE", output_dtype == "bfloat16" ? "ushort" : output_dtype},
+        {"LOCAL_SIZE", LOCAL_SIZE_FWD},
+        {"T_TYPE", target_dtype},
+        {"DTYPE_LOWEST", dtype_lowest},
```

Comment:
You can use log(0) or -INFINITY inside kernels code. This #define flag is unnecessary.

Reply:
Resolved